### ISSUE TITLE
429 Too many requests or 429 API Rate limit exceeded

### ISSUE THREAD

**Issue by chasebolt** [digitalocean/terraform-provider-digitalocean#546]
### Terraform Version
```
Terraform v0.13.5
+ provider registry.terraform.io/-/aws v3.22.0
+ provider registry.terraform.io/-/digitalocean v2.3.0
+ provider registry.terraform.io/digitalocean/digitalocean v2.3.0
+ provider registry.terraform.io/hashicorp/aws v3.22.0
```

### Expected Behavior
Terraform to complete successfully

### Actual Behavior
Terraform is getting errors back from DO API stating either Too Many Requests or API Rate limit exceeded.

### Steps to Reproduce
Manage 100 domain records and it will fail while trying to refresh the state.

### Important Factoids
I am managing 10 domains with about 10-20 records each. I have dropped parallelism down to 1 and I still can't get a successful run without hitting one of these two errors. This isn't an absurdly large terraform run, and  I am unable to get past just a handful of state refreshes before it errors out.

**Comment by zeppelinen** [comment]
+1. 
Same problem for me. I need to manage 200+ droplets.

**Comment by lfarnell** [comment]
Based on the Docs
```
Requests through the API are rate limited per OAuth token. Current rate limits:

5,000 requests per hour
250 requests per minute (5% of the hourly total)
```
So one thing you can possibly do is reach out to support and see if they can increase the limit for `requests per minute`. The only other thing you might want to investigate is looking at breaking down the terraform code into a module per domain so that you won't hit the limit. I realize this might not be possible or suitable for your use case by something to think about.

**Comment by joaomnmoreira** [comment]
+1.

Hundreds of domain records (same domain) to manage.

**Comment by ghomem** [comment]
+1

This happens to me white managing a single large domain (~ 150 records)

**Comment by djmaze** [comment]
Same problem here.

> So one thing you can possibly do is reach out to support and see if they can increase the limit for requests per minute.

I just tried that. The answer was simply that they are doing this "to protect the platform". So they won't increase the limits.

I now asked if they could implement at least something like a *burst limit* which would allow to overreach the rate limit for a short amount of time. That would probably help with use cases such as Terraform.

**Comment by atmosx** [comment]
> Same problem here.
> 
> > So one thing you can possibly do is reach out to support and see if they can increase the limit for requests per minute.
> 
> I just tried that. The answer was simply that they are doing this "to protect the platform". So they won't increase the limits.
> 
> I now asked if they could implement at least something like a _burst limit_ which would allow to overreach the rate limit for a short amount of time. That would probably help with use cases such as Terraform.

I'm the only one having this problem in a team of three. My coworkers can `plan`. I have issued tokens that are used by applications, I wonder if they're using the API and I'm hitting a _global_ limit.

Since @djmaze and me are hit by this at the same point in time, I wonder if there's something else going on.

**Comment by gergelypolonkai** [comment]
Today this hit me for the first API request while refreshing terraform state, without me doing anything on DO except a Kubernetes “login”.

I saw @tback creating a possible fix for it; any chance it will be released soon? (The last release is from two days before the fix landed)

**Comment by Lavode** [comment]
We too regularly run into the per-minute rate-limiting of API calls. As such we too would greatly support a solution to this. Be it in the form of an option for increased rate limits on DO's side, the Terraform provider automatically retrying failed calls, or even just adding support for rate limiting to the provider. We'd much prefer a deployment taking a few minutes longer, over it potentially failing.

Some background on our use-case, should it help: We regularly spawn on the order of 100 VMs for a few hours at a time, in order to profile distributed applications of ours. These deployments are automated with Terraform. Each VM we provision will lead to multiple API calls, as we also set up DNS records for each, assign it to a project, and so on. 

As we regularly get rate limited this then requires multiple calls to `terraform apply` - spaced apart a few minutes each time - for the stack to be deployed fully. This is inconvenient when doing it manually, and a nightmare in CI.

**Comment by tback** [comment]
I just installed the provider manually with a provider override and it worked for me as long as it took me to migrate away from DO.

**Comment by runako** [comment]
+1 Would pay extra to not have to worry about rate-limiting the API, which I use to pay provision resources and therefore pay Digital Ocean more money. If there was an option to pay $X/mo for a much higher/unlimited API, I would choose that option at this moment. My other option is to spend time building rate-limiting into my side of the app, which is going to take way more of my time than $X/mo.

**Comment by benyanke** [comment]
I have similarly filed support requests, as my modestly sized terraform stack is encountering these issues. Unfortunately their own terraform provider isn't usable on their platform.

Perhaps if backoff was added to the provider it would slow things down but not impossibly fail?

**Comment by DanielHLelis** [comment]
I've hit the same problems of rate limits (to the point at which I couldn't even run a single plan anymore) and opened a PR with proposed changes that fixes/mitigate this issue. If anyone would like to take a look at the fork: [`DanielHLelis/terraform-provider-digitalocean-ratelimit`](https://github.com/DanielHLelis/terraform-provider-digitalocean-ratelimit/tree/feature/godo-rate-limiter-and-http-retryable).
It basically limits the number of requests per second and/or uses the retryable HTTP client from Hashicorp to retry the request after an error (like the 422).

To try it you just need to install it and do the override as explained in the `CONTRIBUTING.md`.

Hoping to get it into mainstream soon.

**Comment by benyanke** [comment]
Can you link the PR here too?

**Comment by DanielHLelis** [comment]
> Can you link the PR here too?

Here: [#967](https://github.com/digitalocean/terraform-provider-digitalocean/pull/967)

**Comment by benyanke** [comment]
Updating here to say that I spent over a week going back and forth with support, and they were not at all helpful. Making me wonder if I should move my DNS elsewhere, if DNS + IAC isn't going to be well supported.

**Comment by andrewsomething** [comment]
We've just released version 2.28.0 of this provider. It  adds experimental support for automatically retrying requests that fail with 429 or 500-level response codes. It can be enabled by setting the `DIGITALOCEAN_HTTP_RETRY_MAX` environment variable or the `http_retry_max` argument in the provider configuration.

Additionally, it adds support for configuring client-side rate-limiting to enforce quality of service. It can be enabled by setting the `DIGITALOCEAN_REQUESTS_PER_SECOND` environment variable or the `requests_per_second` argument in the provider configuration.

Please let us know if you have any feedback on this functionality. We will be looking to enable it by default in a future release.

Thanks to @DanielHLelis for working with us on this!

**Comment by benyanke** [comment]
Thanks! This seems to be working initially - I'll do more testing and report back.

**Comment by madalozzo** [comment]
That's amazing @DanielHLelis! I've run the last release with 'requests_per_second' parameter and it  appears to be working very well. I think that now I can merge all my terraform in one eheh

**Comment by jonathanheilmann** [comment]
I reached the rate limit too, after setting `requests_per_second` the error has gone. Looks like it's doing its job well :)

**Comment by andrewsomething** [comment]
With the recently released version 2.30.0 of this provider, we have now retries by default. Setting the `DIGITALOCEAN_HTTP_RETRY_MAX` environment variable or the `http_retry_max` argument in the provider configuration to `0` will disable this behavior.
